frame/vesting/src/lib.rs

```diff
@@ -302,72 +409,273 @@ pub mod pallet {
+				return Ok(());
+			};
+			let who = ensure_signed(origin)?;
+			let schedule1_index = schedule1_index as usize;
+			let schedule2_index = schedule2_index as usize;
+			let vesting = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+			let len = vesting.len();
+			ensure!(
+				schedule1_index < len && schedule2_index < len,
+				Error::<T>::ScheduleIndexOutOfBounds
+			);
+
+			// The schedule index is based off of the schedule ordering prior to filtering out any
+			// schedules that may be ending at this block.
+			let schedule1 = vesting[schedule1_index];
+			let schedule2 = vesting[schedule2_index];
+			let filter = Filter::Two((schedule1_index, schedule2_index));
+
+			// The length of vesting decreases by 2 here since we filter out 2 schedules. Thus we know
+			// below that we can safely insert the new merged schedule.
+			let maybe_vesting = Self::update_lock_and_schedules(who.clone(), vesting, filter);
+
+			// We can't fail from here on because we have potentially removed two schedules.
 
-			Self::add_vesting_schedule(&target, schedule.locked, schedule.per_block, schedule.starting_block)
-				.expect("user does not have an existing vesting schedule; q.e.d.");
+			let now = <frame_system::Pallet<T>>::block_number();
+			if let Some(s) = Self::merge_vesting_info(now, schedule1, schedule2) {
+				let mut vesting = maybe_vesting.unwrap_or_default();
+				// It shouldn't be possible for this to fail because we removed 2 schedules above.
+				ensure!(vesting.try_push(s).is_ok(), Error::<T>::AtMaxVestingSchedules);
+				Self::deposit_event(Event::<T>::VestingMergeSuccess(
+					s.locked(),
+					s.per_block(),
+					s.starting_block(),
+				));
+				Vesting::<T>::insert(&who, vesting);
+			} else if maybe_vesting.is_some() {
+				Vesting::<T>::insert(&who, maybe_vesting.unwrap());
+			} else {
+				Vesting::<T>::remove(&who);
+			}
 
 			Ok(())
 		}
 	}
 }
 
 impl<T: Config> Pallet<T> {
+	// Create a new `VestingInfo`, based off of two other `VestingInfo`s.
+	// NOTE: We assume both schedules have been vested up through the current block.
+	fn merge_vesting_info(
+		now: T::BlockNumber,
+		schedule1: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+		schedule2: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+	) -> Option<VestingInfo<BalanceOf<T>, T::BlockNumber>> {
+		let schedule1_ending_block = schedule1.ending_block::<T::BlockNumberToBalance>();
+		let schedule2_ending_block = schedule2.ending_block::<T::BlockNumberToBalance>();
+		let now_as_balance = T::BlockNumberToBalance::convert(now);
+
+		if schedule1_ending_block <= now_as_balance && schedule2_ending_block <= now_as_balance {
+			// If both schedules hav ended, we don't merge and exit early.
+			return None;
+		} else if schedule1_ending_block <= now_as_balance {
+			// If one schedule has ended, we treat the one that has not ended as the new
+			// merged schedule.
+			return Some(schedule2);
+		} else if schedule2_ending_block <= now_as_balance {
+			return Some(schedule1);
+		}
```

Comment:
We can eliminate some comparison operations and IMO clarify the logic by rewriting as a `match`:

```suggestion
		match (schedule1_ending_block <= now_as_balance, schedule2_ending_block <= now_as_balance) {
			// If both schedules hav ended, we don't merge and exit early.
			(true, true) => return None,
			// If one schedule has ended, we treat the one that has not ended as the new
			// merged schedule.
			(true, false) => return Some(schedule2),
			(false, true) => return Some(schedule1),
			_ => {},
		}
```